### ISSUE TITLE
5: Stochastic Optimization for Large-Scale Inference

### ISSUE THREAD

**Issue by jwintz** [IGoR-immuno-stack/IGoR#11]
# Task 5: Stochastic Optimization for Large-Scale Inference

## Current State

IGoR uses **batch Expectation-Maximization (EM)** for model inference. The entire dataset is processed in each iteration before updating model parameters.

### Current Implementation

**In [src/igor/Core/GenModel.cpp](src/igor/Core/GenModel.cpp) (lines 47-55):**
```cpp
bool GenModel::infer_model(
    const vector<tuple<int,string,unordered_map<Gene_class , vector<Alignment_data>>>>& sequences,
    const int iterations,
    const string path,
    bool fast_iter/*=true*/,
    double likelihood_threshold/*=1e-25 by default*/,
    bool viterbi_like/*=false*/,
    double proba_threshold_factor/*=0.001 by default*/,
    double mean_number_seq_err_thresh /*= INFINITY by default*/)
```

**Main inference loop structure (lines 127-300):**
```cpp
// Loop over iterations
while(iteration_accomplished!=iterations){
    Model_marginals new_marginals = Model_marginals(model_parms);
    // ... initialization ...
    
    // Process EVERY sequence in the dataset
    #pragma omp parallel for schedule(dynamic)
    for(auto seq_it = sequences.begin(); seq_it < sequences.end(); ++seq_it){
        // Evaluate sequence and accumulate marginals
        first_event->iterate(...);
        // Add to single_thread_marginals
        single_thread_marginals += single_seq_marginals;
    }
    
    // Merge all thread marginals
    new_marginals += single_thread_marginals;
    
    // Normalize and update model (M-step)
    new_marginals.normalize(...);
    this->model_marginals = new_marginals;
    ++iteration_accomplished;
}
```

### Current Characteristics

| Aspect | Current Behavior |
|--------|-----------------|
| **Update Frequency** | Once per full dataset pass |
| **Memory** | All sequences + all marginals in RAM |
| **Parallelism** | Per-sequence within iteration |
| **Convergence** | Guaranteed but slow for large N |
| **Learning Rate** | Implicit (full batch = LR of 1.0) |

### Performance Benchmarks (Estimated)

| Dataset Size | Time per Iteration | Memory Usage | Total Time (10 iter) |
|--------------|-------------------|--------------|----------------------|
| 10,000 seqs  | ~5 min            | ~1 GB        | ~50 min              |
| 100,000 seqs | ~50 min           | ~10 GB       | ~8 hrs               |
| 1,000,000 seqs | ~8 hrs          | ~100 GB      | ~3 days              |

## Problem

1. **Scalability**: Modern repertoire datasets contain millions of sequences
2. **Convergence Speed**: Full-batch EM requires many passes to converge
3. **Resource Inefficiency**: Each iteration processes the entire dataset before any update
4. **No Early Stopping**: Cannot checkpoint or evaluate convergence mid-iteration
5. **Memory Bottleneck**: Entire dataset must fit in memory (see Task 6)

### Mathematical Background

Batch EM update rule:
$$\theta_{t+1} = \arg\max_\theta \mathbb{E}_{z|x,\theta_t}[\log P(x,z|\theta)]$$

This requires computing expectations over ALL sequences before updating $\theta$.

Stochastic EM update rule:
$$\theta_{t+1} = \theta_t + \eta_t \cdot \left( \mathbb{E}_{z|x_B,\theta_t}[\log P(x_B,z|\theta)] - \theta_t \right)$$

Where $x_B$ is a mini-batch and $\eta_t$ is a decaying learning rate.

## Desired State

1. **Mini-batch Processing**: Update model after each mini-batch of sequences
2. **Online Learning**: More frequent updates → faster convergence in wall-clock time
3. **Scalability**: Process datasets of arbitrary size
4. **Configurable Learning Rate**: Support various scheduling strategies
5. **Convergence Monitoring**: Track likelihood on validation set
6. **Checkpointing**: Save intermediate models during training

### Target Performance

| Dataset Size | Time to Convergence | Memory Usage |
|--------------|-------------------|--------------|
| 10,000 seqs  | ~10-20 min        | ~500 MB      |
| 100,000 seqs | ~30-60 min        | ~500 MB      |
| 1,000,000 seqs | ~2-4 hrs        | ~500 MB      |

### Stochastic EM Algorithm

```
Algorithm: Stochastic Expectation-Maximization

Input: Dataset D, initial θ₀, batch_size b, learning_rate_schedule η(t)
Output: Learned parameters θ*

1. Initialize t = 0
2. While not converged:
   a. Sample mini-batch B_t of size b from D (with or without replacement)
   b. E-step: Compute sufficient statistics S(B_t, θ_t) for batch
   c. M-step: θ_{t+1} = θ_t + η(t) × (normalize(S(B_t, θ_t)) - θ_t)
   d. t = t + 1
   e. If t mod eval_frequency == 0: evaluate on validation set
3. Return θ_t
```

#### StochasticConfig

```cpp
// New file: src/igor/Core/StochasticInference.h

struct StochasticConfig {
    // Batch settings
    size_t batch_size = 1000;
    bool sample_with_replacement = true;
    
    // Learning rate
    double initial_learning_rate = 1.0;
    double learning_rate_decay = 0.01;      // η(t) = η₀ / (1 + decay × t)
    double min_learning_rate = 0.01;
    
    // Convergence
    size_t min_iterations = 10;
    size_t max_iterations = 10000;
    double convergence_threshold = 1e-4;
    size_t convergence_window = 10;         // Check last N iterations
    
    // Advanced
    bool use_momentum = false;
    double momentum = 0.9;
    bool use_adam = false;
    double adam_beta1 = 0.9;
    double adam_beta2 = 0.999;
    
    // Checkpointing
    size_t checkpoint_frequency = 100;      // Save every N batches
    size_t eval_frequency = 50;             // Evaluate on validation every N batches
    double validation_fraction = 0.1;       // Hold out 10% for validation
    
    // Reproducibility
    uint64_t random_seed = 0;               // 0 = use time
};
```

#### Learning Rate Schedules

```cpp
enum class LearningRateSchedule {
    CONSTANT,           // η(t) = η₀
    INVERSE_DECAY,      // η(t) = η₀ / (1 + decay × t)
    EXPONENTIAL_DECAY,  // η(t) = η₀ × exp(-decay × t)
    STEP_DECAY,         // η(t) = η₀ × 0.5^(t / step_size)
    COSINE_ANNEALING,   // η(t) = η_min + 0.5(η₀ - η_min)(1 + cos(πt/T))
    WARMUP_LINEAR       // Linear warmup then linear decay
};

class LearningRateScheduler {
public:
    LearningRateScheduler(LearningRateSchedule schedule, 
                          const StochasticConfig& config);
    
    double get_learning_rate(size_t iteration) const;
    void set_max_iterations(size_t max_iter);
    
private:
    LearningRateSchedule schedule_;
    double initial_lr_;
    double min_lr_;
    double decay_;
    size_t warmup_steps_;
    size_t max_iterations_;
};
```

#### StochasticInference Class

```cpp
class StochasticInference {
public:
    StochasticInference(
        Model_Parms& model_parms,
        Model_marginals& model_marginals,
        const StochasticConfig& config = StochasticConfig());
    
    // Main inference methods
    bool infer(
        SequenceDataSource& data_source,  // See Task 6
        const std::string& output_path);
    
    // With progress callback
    bool infer(
        SequenceDataSource& data_source,
        const std::string& output_path,
        std::function<void(const InferenceProgress&)> callback);
    
    // Resume from checkpoint
    bool resume(
        const std::string& checkpoint_path,
        SequenceDataSource& data_source,
        const std::string& output_path);
    
    // Getters for monitoring
    size_t current_iteration() const;
    double current_train_likelihood() const;
    double current_validation_likelihood() const;
    double effective_learning_rate() const;
    
    // Manual control
    void step();  // Perform one batch update
    bool has_converged() const;
    void save_checkpoint(const std::string& path) const;
    void load_checkpoint(const std::string& path);
    
private:
    // Core methods
    double process_batch(
        const std::vector<SequenceData>& batch,
        Model_marginals& batch_marginals);
    
    void update_model(
        const Model_marginals& batch_marginals,
        double learning_rate);
    
    void update_model_with_momentum(
        const Model_marginals& batch_marginals,
        double learning_rate);
    
    bool check_convergence();
    
    // State
    Model_Parms& model_parms_;
    Model_marginals& model_marginals_;
    StochasticConfig config_;
    LearningRateScheduler lr_scheduler_;
    
    // Iteration state
    size_t current_iteration_;
    std::vector<double> train_likelihood_history_;
    std::vector<double> validation_likelihood_history_;
    
    // Momentum state (if enabled)
    Model_marginals momentum_marginals_;
    
    // Adam state (if enabled)
    Model_marginals adam_m_;  // First moment
    Model_marginals adam_v_;  // Second moment
    
    // Random state
    std::mt19937_64 rng_;
};
```

#### Progress Reporting

```cpp
struct InferenceProgress {
    size_t iteration;
    size_t total_sequences_processed;
    double train_likelihood;
    double validation_likelihood;
    double learning_rate;
    double elapsed_seconds;
    double estimated_remaining_seconds;
    bool converged;
};
```

### Integration with Existing GenModel

```cpp
// Update GenModel.h to support stochastic inference

class GenModel {
public:
    // Existing batch inference (unchanged)
    bool infer_model(const vector<...>& sequences, int iterations, ...);
    
    // New stochastic inference
    bool infer_model_stochastic(
        SequenceDataSource& data_source,
        const StochasticConfig& config,
        const std::string& path);
    
    // Resume training
    bool resume_inference(
        const std::string& checkpoint_path,
        SequenceDataSource& data_source,
        const std::string& output_path);
    
private:
    std::unique_ptr<StochasticInference> stochastic_inference_;
};
```

### Model Marginals Update with Learning Rate

```cpp
// In Model_marginals.cpp

void Model_marginals::update_with_learning_rate(
    const Model_marginals& batch_marginals,
    double learning_rate) 
{
    // θ_{t+1} = θ_t + η × (batch_marginals - θ_t)
    // Equivalent to: θ_{t+1} = (1 - η) × θ_t + η × batch_marginals
    
    const size_t len = this->get_length();
    for (size_t i = 0; i < len; ++i) {
        this->marginal_array_smart_p[i] = 
            (1.0 - learning_rate) * this->marginal_array_smart_p[i] +
            learning_rate * batch_marginals.marginal_array_smart_p[i];
    }
}

void Model_marginals::update_with_momentum(
    const Model_marginals& batch_marginals,
    Model_marginals& momentum,
    double learning_rate,
    double momentum_coef)
{
    // v_{t+1} = momentum × v_t + η × (batch_marginals - θ_t)
    // θ_{t+1} = θ_t + v_{t+1}
    
    const size_t len = this->get_length();
    for (size_t i = 0; i < len; ++i) {
        double gradient = batch_marginals.marginal_array_smart_p[i] - 
                         this->marginal_array_smart_p[i];
        momentum.marginal_array_smart_p[i] = 
            momentum_coef * momentum.marginal_array_smart_p[i] +
            learning_rate * gradient;
        this->marginal_array_smart_p[i] += momentum.marginal_array_smart_p[i];
    }
}
```

### Subtask 5.1: Design Stochastic EM Algorithm
- [ ] Define mathematical formulation for IGoR's model
- [ ] Analyze sufficient statistics structure
- [ ] Design learning rate schedule options
- [ ] Document convergence guarantees and assumptions

### Subtask 5.2: Implement StochasticConfig
- [ ] Create configuration structure
- [ ] Implement validation of parameters
- [ ] Support JSON/YAML configuration loading
- [ ] Unit tests for configuration

### Subtask 5.3: Implement LearningRateScheduler
- [ ] Implement all schedule types
- [ ] Add warmup support
- [ ] Add cosine annealing with restarts
- [ ] Unit tests for each schedule

### Subtask 5.4: Implement StochasticInference Core
- [ ] Implement `process_batch()` method
- [ ] Implement `update_model()` with learning rate
- [ ] Implement batch sampling (with/without replacement)
- [ ] Handle thread safety for parallel batch processing

### Subtask 5.5: Implement Momentum and Adam
- [ ] Implement momentum update rule
- [ ] Implement Adam optimizer
- [ ] Handle marginal normalization constraints
- [ ] Benchmark optimizer performance

### Subtask 5.6: Implement Convergence Detection
- [ ] Track likelihood history
- [ ] Implement window-based convergence check
- [ ] Implement early stopping
- [ ] Implement patience mechanism

### Subtask 5.7: Implement Checkpointing
- [ ] Design checkpoint file format
- [ ] Implement `save_checkpoint()`
- [ ] Implement `load_checkpoint()`
- [ ] Include optimizer state (momentum, Adam)

### Subtask 5.8: Implement Validation Evaluation
- [ ] Split data into train/validation
- [ ] Evaluate on validation set periodically
- [ ] Track validation likelihood for early stopping
- [ ] Avoid overfitting detection

### Subtask 5.9: Update GenModel Interface
- [ ] Add `infer_model_stochastic()` method
- [ ] Add `resume_inference()` method
- [ ] Maintain backward compatibility
- [ ] Update documentation

### Subtask 5.10: Command Line Interface
- [ ] Add `--stochastic` flag
- [ ] Add `--batch-size N` option
- [ ] Add `--learning-rate X` option
- [ ] Add `--lr-schedule TYPE` option
- [ ] Add `--checkpoint PATH` option
- [ ] Update help text

### Subtask 5.11: Logging and Monitoring
- [ ] Create inference log file format
- [ ] Log per-batch statistics
- [ ] Log learning rate over time
- [ ] Support TensorBoard-style output (optional)

### Subtask 5.12: Integration Testing
- [ ] Test convergence on small datasets
- [ ] Compare final likelihood to batch EM
- [ ] Test checkpoint/resume functionality
- [ ] Benchmark wall-clock speedup

### Subtask 5.13: Performance Optimization
- [ ] Profile batch processing
- [ ] Optimize memory allocation
- [ ] Parallelize batch processing
- [ ] Benchmark scaling with batch size

### Basic Stochastic Inference

```shell
igor -set_wd /path/to/workdir \
     -batch large_dataset \
     -species human -chain beta \
     --stochastic \
     --batch-size 1000 \
     --max-iterations 5000 \
     -infer

# Expected output:
# Initializing stochastic inference...
# Batch size: 1000, Initial learning rate: 1.0
# Learning rate schedule: inverse_decay (decay=0.01)
# 
# Iteration 100: train_ll=-45.23, val_ll=-45.31, lr=0.50
# Iteration 200: train_ll=-42.15, val_ll=-42.25, lr=0.33
# Iteration 300: train_ll=-40.87, val_ll=-40.92, lr=0.25
# ...
# Converged at iteration 2341
# Final log-likelihood: -38.42
# Total time: 45 minutes
```

### With Custom Learning Rate

```shell
igor -set_wd /path/to/workdir \
     -batch large_dataset \
     -species human -chain beta \
     --stochastic \
     --batch-size 2000 \
     --learning-rate 0.5 \
     --lr-schedule cosine \
     --lr-min 0.01 \
     --max-iterations 10000 \
     -infer
```

### With Checkpointing

```shell
# Start training with checkpoints
igor -set_wd /path/to/workdir \
     -batch large_dataset \
     --stochastic \
     --checkpoint-freq 500 \
     -infer

# Resume from checkpoint
igor -set_wd /path/to/workdir \
     -batch large_dataset \
     --resume /path/to/checkpoint_iter_5000.ckpt \
     -infer
```

### With Momentum

```shell
igor -set_wd /path/to/workdir \
     -batch large_dataset \
     --stochastic \
     --batch-size 500 \
     --use-momentum \
     --momentum 0.9 \
     -infer
```

### Output Files

```
workdir/
├── stochastic_inference_log.csv
│   # iteration,train_likelihood,val_likelihood,learning_rate,time
│   1,-50.23,-50.31,1.0,0.5
│   2,-49.87,-49.92,0.99,1.0
│   ...
├── checkpoints/
│   ├── checkpoint_iter_500.ckpt
│   ├── checkpoint_iter_1000.ckpt
│   └── ...
├── final_marginals.txt
└── final_parms.txt
```

## Success Criteria

1. **Correctness**: Final likelihood comparable to batch EM (within 1%)
2. **Speed**: 2-10x faster wall-clock time to convergence on large datasets
3. **Memory**: Constant memory usage regardless of dataset size
4. **Stability**: Convergence achieved reliably across different seeds
5. **Usability**: Simple command-line interface
6. **Robustness**: Checkpoint/resume works correctly

### Sufficient Statistics for IGoR

For the EM algorithm, sufficient statistics are:
- **Gene choice events**: Count of each gene selection weighted by posterior
- **Deletion events**: Count of each deletion length weighted by posterior
- **Insertion events**: Count of each insertion length weighted by posterior
- **Dinucleotide Markov**: Count of dinucleotide transitions weighted by posterior

These are exactly what's accumulated in `Model_marginals`.

### Learning Rate Theory

For stochastic EM to converge, learning rate must satisfy:
$$\sum_{t=1}^{\infty} \eta_t = \infty \quad \text{and} \quad \sum_{t=1}^{\infty} \eta_t^2 < \infty$$

The schedule $\eta_t = \eta_0 / (1 + \text{decay} \times t)$ satisfies these conditions.

### Variance Reduction

Mini-batch averaging reduces gradient variance:
$$\text{Var}[\nabla_\theta L_B] = \frac{\text{Var}[\nabla_\theta L_1]}{b}$$

where $b$ is batch size. Larger batches → lower variance → can use higher learning rate.
